File: mcps/unity/server/src/invocation-log.ts
/**
 * Server-side per-invocation log lines. Pairs with the Editor's InvocationLog
 * so each tool call has matching `[unity-mcp] <id> <tool>` lines on both sides
 * of the bridge. The correlation ID is shared end-to-end (it's the same UUID
 * the Bridge writes into the request envelope and the Editor logs against).
 *
 * Verbosity controlled by the UNITY_MCP_LOG_LEVEL env var:
 *   - unset / "all"          → emit start + ok/err (default)
 *   - "errors-only" / "err"  → emit err only
 */
const PREFIX = "[unity-mcp]";

function trimId(correlationId: string): string {
  if (!correlationId) return "(no-id)";
  return correlationId.length > 8 ? correlationId.slice(0, 8) : correlationId;
}

function errorsOnly(): boolean {
  const v = process.env.UNITY_MCP_LOG_LEVEL;
  if (!v) return false;
  const lower = v.toLowerCase();
  return lower === "errors-only" || lower === "err";
}

function firstLine(message: string): string {
  if (!message) return "";
  const nl = message.search(/[\r\n]/);
  return nl >= 0 ? message.slice(0, nl) : message;
}

export interface InvocationLogEntry {
  ok(): void;
  err(code: string, message: string): void;
}

export function startInvocation(correlationId: string, tool: string): InvocationLogEntry {
  const start = process.hrtime.bigint();
  if (!errorsOnly()) {
    console.error(`${PREFIX} ${trimId(correlationId)} ${tool} start`);
  }
  return {
    ok() {
      if (errorsOnly()) return;
      const ms = Number((process.hrtime.bigint() - start) / 1_000_000n);
      console.error(`${PREFIX} ${trimId(correlationId)} ${tool} ok ${ms}ms`);
    },
    err(code: string, message: string) {
      const ms = Number((process.hrtime.bigint() - start) / 1_000_000n);
      // Errors always emit, regardless of verbosity.
      console.error(`${PREFIX} ${trimId(correlationId)} ${tool} err ${code} ${ms}ms ${firstLine(message)}`);
    },
  };
}

File: mcps/unity/unity-package/Editor/Bridge/InvocationLog.cs
using System;
using System.Diagnostics;
using Debug = UnityEngine.Debug;

namespace UnityMcp.Editor.Bridge
{
    /// <summary>
    /// Structured per-tool-invocation log lines. Pairs with the server-side
    /// invocation-log so every tool call produces matching `[UnityMCP] &lt;id&gt; &lt;tool&gt;`
    /// entries on both halves of the bridge — searchable by correlation ID end-to-end.
    ///
    /// Verbosity controlled by the <c>UNITY_MCP_LOG_LEVEL</c> env var:
    ///   - unset / "all"          → emit start + ok/err (default)
    ///   - "errors-only" / "err"  → emit err only
    /// </summary>
    internal static class InvocationLog
    {
        private const string EnvVar = "UNITY_MCP_LOG_LEVEL";
        private const string Prefix = "[UnityMCP]";

        public static Stopwatch Start(string correlationId, string tool)
        {
            if (!ErrorsOnly())
            {
                Debug.Log($"{Prefix} {Trim(correlationId)} {tool} start");
            }
            return Stopwatch.StartNew();
        }

        public static void Ok(string correlationId, string tool, Stopwatch sw)
        {
            if (ErrorsOnly()) return;
            sw?.Stop();
            long ms = sw?.ElapsedMilliseconds ?? -1;
            Debug.Log($"{Prefix} {Trim(correlationId)} {tool} ok {ms}ms");
        }

        public static void Err(string correlationId, string tool, Stopwatch sw, string code, string message)
        {
            sw?.Stop();
            long ms = sw?.ElapsedMilliseconds ?? -1;
            // Errors always emit, regardless of verbosity. First line of the message only —
            // multi-line stacks pollute the structured-log shape.
            string firstLine = FirstLine(message);
            Debug.Log($"{Prefix} {Trim(correlationId)} {tool} err {code} {ms}ms {firstLine}");
        }

        private static bool ErrorsOnly()
        {
            try
            {
                var v = Environment.GetEnvironmentVariable(EnvVar);
                return v != null && (v.Equals("errors-only", StringComparison.OrdinalIgnoreCase)
                                  || v.Equals("err", StringComparison.OrdinalIgnoreCase));
            }
            catch
            {
                return false;
            }
        }

        private static string Trim(string correlationId)
        {
            if (string.IsNullOrEmpty(correlationId)) return "(no-id)";
            // Full UUIDs are noisy in the console; first 8 chars stay greppable.
            return correlationId.Length > 8 ? correlationId.Substring(0, 8) : correlationId;
        }

        private static string FirstLine(string message)
        {
            if (string.IsNullOrEmpty(message)) return string.Empty;
            int nl = message.IndexOfAny(new[] { '\n', '\r' });
            return nl >= 0 ? message.Substring(0, nl) : message;
        }
    }
}
